src/NHibernate/Engine/ISessionImplementor.cs

```diff
@@ -48,6 +48,41 @@ internal static void AutoFlushIfRequired(this ISessionImplementor implementor, I
 		{
 			(implementor as AbstractSessionImpl)?.AutoFlushIfRequired(querySpaces);
 		}
+
+		/// <summary>
+		/// Switch the session current cache mode.
+		/// </summary>
+		/// <param name="session">The session for which the cache mode has to be switched.</param>
+		/// <param name="cacheMode">The desired cache mode. <see langword="null" /> for not actually switching.</param>
+		/// <returns><see langword="null" /> if no switch is required, otherwise an <see cref="IDisposable"/> which
+		/// dispose will set the session cache mode back to its original value.</returns>
+		internal static IDisposable SwitchCacheMode(this ISessionImplementor session, CacheMode? cacheMode)
+		{
+			if (session == null || !cacheMode.HasValue || cacheMode == session.CacheMode)
```

Comment:
`session == null` check looks unnecessary in internal method. Unless it's a live scenario I think it's actually better to simply allow it to throw NRE on `session.CacheMode` access.

Reply:
If it has to change I would instead throw an `ArgumentNullException`. I do not want to code NRE cases. Anyway that part will go away once merged into `ISessionImplementor`.

Reply:
Ah. Ok. I've missed the fact that you are actually planning to include this method in interface.

Reply:
But then.. this `session == null` check only makes things worse. If there is such case in code - it will be silently ignored now. But when merged to interface - NRE will happen.

Reply:
That is true, but it is done also on `BeginContext` and  `BeginProcess` right above.

Reply:
And is it really required to go in interface? It seems to me that it can leave as internal static extension. It doesn't depend on any specific implementation. So why to clutter interface with something that can be implemented on top of existing members?

Reply:
It is a possibility.